pkg/operators/spoke/controller.go

```diff
@@ -2,28 +2,395 @@ package spoke
+		resourceapply.NewKubeClientHolder(n.kubeClient),
+		controllerContext.Recorder(),
+		func(name string) ([]byte, error) {
+			return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join("", name)), config).Data, nil
+		},
+		"manifests/spoke/spoke-clusterrolebinding.yaml",
+		"manifests/spoke/spoke-serviceaccount.yaml",
+	)
+	errs := []error{}
+	for _, result := range resourceResults {
+		if result.Error != nil {
+			errs = append(errs, fmt.Errorf("%q (%T): %v", result.File, result.Type, result.Error))
+		}
+	}
+
+	if len(errs) > 0 {
+		appleErros := operatorhelpers.NewMultiLineAggregate(errs)
+		helpers.SetNucleusCondition(&spokeCore.Status.Conditions, nucleusapiv1.StatusCondition{
+			Type:    spokeCoreApplied,
+			Status:  metav1.ConditionFalse,
+			Reason:  "SpokeCoreApplyFailed",
+			Message: appleErros.Error(),
+		})
+		helpers.UpdateNucleusSpokeStatus(
+			ctx, n.nucleusClient, spokeCoreName, helpers.UpdateNucleusSpokeConditionFn(spokeCore.Status.Conditions...))
+		return appleErros
+	}
+
+	// Create hub config secret
+	hubSecret, err := n.kubeClient.CoreV1().Secrets(config.SpokeCoreNamespace).Get(
+		ctx, hubKubeConfigSecret, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// Craete an empty secret with placeholder
+		hubSecret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      hubKubeConfigSecret,
+				Namespace: config.SpokeCoreNamespace,
+			},
+			Data: map[string][]byte{"placeholder": []byte("placeholder")},
+		}
+		hubSecret, err = n.kubeClient.CoreV1().Secrets(config.SpokeCoreNamespace).Create(ctx, hubSecret, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if err != nil {
+		helpers.SetNucleusCondition(&spokeCore.Status.Conditions, nucleusapiv1.StatusCondition{
+			Type:    spokeCoreApplied,
+			Status:  metav1.ConditionFalse,
+			Reason:  "SpokeCoreApplyFailed",
+			Message: fmt.Sprintf("Failed to get hub kubeconfig secret with error %v", err),
+		})
+		helpers.UpdateNucleusSpokeStatus(
+			ctx, n.nucleusClient, spokeCoreName, helpers.UpdateNucleusSpokeConditionFn(spokeCore.Status.Conditions...))
+		return err
+	}
+
+	// Deploy registration agent
+	generation, err := n.applyDeployment(
+		config, "manifests/spoke/spoke-registration-deployment.yaml", n.registrationGeneration, controllerContext)
+	if err != nil {
+		helpers.SetNucleusCondition(&spokeCore.Status.Conditions, nucleusapiv1.StatusCondition{
+			Type:    spokeCoreApplied,
+			Status:  metav1.ConditionFalse,
+			Reason:  "SpokeCoreApplyFailed",
+			Message: fmt.Sprintf("Failed to deploy registration deployment with error %v", err),
+		})
+		helpers.UpdateNucleusSpokeStatus(
+			ctx, n.nucleusClient, spokeCoreName, helpers.UpdateNucleusSpokeConditionFn(spokeCore.Status.Conditions...))
+		return err
+	}
+	n.registrationGeneration = generation
+
+	// If cluster name is empty, read cluster name from hub config secret
+	if config.ClusterName == "" {
+		clusterName := hubSecret.Data["cluster-name"]
```

Comment:
I'm asking myself as I read this whether we should formalize an API for what's in this secret.